src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/prisonperson/MigrationService.kt

```diff
@@ -21,51 +23,57 @@ import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.Migration
 import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.MigrationType
 import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.NomisApiService
 
-abstract class PrisonPersonMigrationService(
+@Service("prisonPersonMigrationService")
+class MigrationService(
   queueService: MigrationQueueService,
   private val nomisService: NomisApiService,
-  private val prisonPersonMappingService: PrisonPersonMappingApiService,
+  @Qualifier("prisonPersonMappingApiService") private val mappingApiService: MappingApiService,
   migrationHistoryService: MigrationHistoryService,
   telemetryClient: TelemetryClient,
   auditService: AuditService,
   @Value("\${page.size:1000}") pageSize: Long,
   @Value("\${complete-check.delay-seconds}") completeCheckDelaySeconds: Int,
   @Value("\${complete-check.count}") completeCheckCount: Int,
-) : MigrationService<PrisonPersonMigrationFilter, PrisonPersonMigrationRequest, PrisonerPhysicalAttributesResponse, PrisonPersonMigrationMappingRequest>(
+  private val entityMigrators: List<EntityMigrator<*, *>>,
+) : MigrationService<MigrationFilter, MigrationRequest, PrisonerPhysicalAttributesResponse, PrisonPersonMigrationMappingRequest>(
   queueService = queueService,
   auditService = auditService,
   migrationHistoryService = migrationHistoryService,
-  mappingService = prisonPersonMappingService,
+  mappingService = mappingApiService,
   telemetryClient = telemetryClient,
   migrationType = MigrationType.PRISONPERSON,
   pageSize = pageSize,
   completeCheckDelaySeconds = completeCheckDelaySeconds,
   completeCheckCount = completeCheckCount,
 ) {
-  abstract suspend fun migrateEntity(offenderNo: String): List<Long>
+  private suspend fun migrateEntity(
+    offenderNo: String,
+    migrationType: PrisonPersonMigrationMappingRequest.MigrationType,
+  ): List<Long> = entityMigrators.find { it.supportsType(migrationType.javaClass) }?.migrateEntity(offenderNo)
+    ?: throw IllegalArgumentException("No entity migrator found for migration type $migrationType")
```

Comment:
For the type specific entity migrations I've had to resort to the visitor pattern based on the MigrationType. It got very messy with generics so dropped that, but there might be a better way than this 🤔